### PR TITLE
fix(MMU_BLAKE): add test + debug

### LINKED ISSUE #1516: BlakeSubsection null
in test CALLBlake2f.json, we try to create a BlakeSubsection but it fails because there isn't enough gaz:

 blakeSuccess = wellFormedF && sufficientGas = true && false = false;
 
 but then the condition:
 
 checkArgument(blake2fParamsOobCall.isRamSuccess() == blakeSuccess);
 
 
is not verified, throw an exception. This lead to the blakeSubsection not being  instantiated and later to an NPE.

Is the check correct? Or should we set success to false in blake2fParamsOobCall = new Blake2fParamsOobCall(calleeGas); ?

### LINKED ISSUE #1517: wrong calculation for numberMmioInstructions
numberMmioInstructions is computed as:

  public int numberMmioInstructions() {
    return totalLeftZeroesInitials + totalNonTrivialInitials + totalRightZeroesInitials;
  }

but it happens it can be bigger. This seems to happen when going through MLoad L89:  public MmuData setMicroInstructions(MmuData mmuData) {

where we set two microinstructions while we have totalNonTrivialInitials 1 only:

Issue is in:

      case MMU_INST_BLAKE -> {
        mmuData = blake.preProcess(mmuData);
        yield blake.setMicroInstructions(mmuData);
      }
      
   blake.preProcess   code set this to one but then we enter, setMicroInstructions we add two of them

### PR DESCRIPTION
**Pull request by letypequividelespoubelles**

### CODE REVIEW

File: arithmetization/src/test/java/net/consensys/linea/zktracer/precompiles/BlakeTests.java [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+            .compile();
+    BytecodeRunner.of(bytecode).run();
+  }
+
+  @Test
+  void notEnoughGasBlakeTest() {
+    final Bytes bytecode =
+        BytecodeCompiler.newProgram()
+            .push(Bytes.fromHexString("0x0badb077")) // value, some random data to hash
+            .push(5) // offset
```
Comment (OlivierBBB) [inline arithmetization/src/test/java/net/consensys/linea/zktracer/precompiles/BlakeTests.java:100]: What is the intent here. You are producing this in memory:

```
<-- 5 bytes -->    <--      32 bytes       -->
00 00 00 00 00  |  00 00 00 ... 00 0b ad b0 77
```

File: arithmetization/src/test/java/net/consensys/linea/zktracer/precompiles/BlakeTests.java [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+            .op(OpCode.MSTORE)
+            .push(10) // value = r * 256 ** 3  for Blake call
+            .push(0) // offset
```
Comment (OlivierBBB) [inline arithmetization/src/test/java/net/consensys/linea/zktracer/precompiles/BlakeTests.java:103]: What is the intent here. You are producing this in memory:
And now

```
<-- 5 bytes -->    <--      32 bytes       -->
0a 00 00 00 00  |  00 00 00 ... 00 0b ad b0 77
```
So you have `r = 0x 0a 00 00 00 = 10 * 256 ** 3` rounds (and gas cost.)

File: arithmetization/src/test/java/net/consensys/linea/zktracer/precompiles/BlakeTests.java [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+            .compile();
+    BytecodeRunner.of(bytecode).run();
+  }
+
+  @Test
+  void wrongFInputTest() {
+    final Bytes bytecode =
+        BytecodeCompiler.newProgram()
+            .push(Bytes.fromHexString("0x0badb077")) // value, some random data to hash
+            .push(5) // offset
```
Comment (OlivierBBB) [inline arithmetization/src/test/java/net/consensys/linea/zktracer/precompiles/BlakeTests.java:77]: Same issue

File: arithmetization/src/test/java/net/consensys/linea/zktracer/precompiles/BlakeTests.java [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+    BytecodeRunner.of(bytecode).run();
+  }
+
+  @Test
+  void basicBlakeTest() {
+    final Bytes bytecode =
+        BytecodeCompiler.newProgram()
+            .push(Bytes.fromHexString("0x0badb077")) // value, some random data to hash
+            .push(5) // offset
+            .op(OpCode.MSTORE)
```
Comment (OlivierBBB) [inline arithmetization/src/test/java/net/consensys/linea/zktracer/precompiles/BlakeTests.java:51]: I really don't understand  the `5` here. Is it to produce some gibberish value ? Don't get me wrong, that'll do the job, it's just surprising.

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/mmu/instructions/Blake.java [hunk 5]
```diff
@@ -148,6 +149,7 @@ public MmuData setMicroInstructions(MmuData mmuData) {
             .sourceLimbOffset(sourceLimbOffsetF)
             .sourceByteOffset(sourceByteOffsetF)
             .targetByteOffset((short) (LLARGE - 1))
+            .targetLimbOffset(1)
```
Comment (OlivierBBB) [inline arithmetization/src/main/java/net/consensys/linea/zktracer/module/mmu/instructions/Blake.java:152]: Nice.

File: arithmetization/src/test/java/net/consensys/linea/zktracer/precompiles/BlakeTests.java [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+  void wrongFInputTest() {
+    final Bytes bytecode =
+        BytecodeCompiler.newProgram()
+            .push(Bytes.fromHexString("0x0badb077")) // value, some random data to hash
+            .push(5) // offset
+            .op(OpCode.MSTORE)
+            .push(2) // value = f for Blake call
+            .push(212) // offset
+            .op(OpCode.MSTORE8)
+            .push(LLARGE * (INDEX_MAX_BLAKE_RESULT + 1)) // return size
```
Comment (OlivierBBB) [inline arithmetization/src/test/java/net/consensys/linea/zktracer/precompiles/BlakeTests.java]: You actually have `BLAKE2f_HASH_OUTPUT_SIZE` in the code elsewhere.

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/imc/mmu/MmuCall.java [hunk 7]
```diff
@@ -587,7 +591,7 @@ public static MmuCall parameterExtractionForBlake(
         .successBit(blakeSuccess)
         .limb1(blakeR)
         .limb2(blakeF)
-        .setBlakeModexp()
+        .setBlakeModexp(blakeSuccess)
```
Comment (OlivierBBB) [inline arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/imc/mmu/MmuCall.java:594]: 👍

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/blake2fmodexpdata/BlakeModexpDataOperation.java [hunk 8]
```diff
@@ -123,9 +124,9 @@ private void traceParameter(Trace trace, UnsignedByte stamp) {
   }
 
   private void traceBlakeResult(Trace trace, UnsignedByte stamp) {
-    final Bytes input = computeBlake2fResult();
+    final Bytes hash = computeBlake2fResult();
```
Comment (OlivierBBB) [inline arithmetization/src/main/java/net/consensys/linea/zktracer/module/blake2fmodexpdata/BlakeModexpDataOperation.java:127]: Was that a bug ? or just overloading the word `input` ?

**Review by OlivierBBB — APPROVED** [review]
LGTM

File: arithmetization/src/test/java/net/consensys/linea/zktracer/precompiles/BlakeTests.java (hunk 1 above)
Comment (letypequividelespoubelles) [reply arithmetization/src/test/java/net/consensys/linea/zktracer/precompiles/BlakeTests.java:100]: yes, I just want to have something (ie  not 0) in the hash input data, so really don't care of what I put inside as soon as we have a non zero between offset 4 and 212.

File: arithmetization/src/test/java/net/consensys/linea/zktracer/precompiles/BlakeTests.java (hunk 2 above)
Comment (letypequividelespoubelles) [reply arithmetization/src/test/java/net/consensys/linea/zktracer/precompiles/BlakeTests.java:103]: yes, I want to produce an OOG (wih the MSTORE8) and just have a non zero input for the hash (same comment as before). It's true we don't care about the input of the hash as we're OOG, but still

File: arithmetization/src/test/java/net/consensys/linea/zktracer/precompiles/BlakeTests.java (hunk 4 above)
Comment (letypequividelespoubelles) [reply arithmetization/src/test/java/net/consensys/linea/zktracer/precompiles/BlakeTests.java:51]: I put a 5 just to be sure it doesn't overlap with the 4 first bytes of the `r`. But yes the value will be leftPadded in the memory, but I don't care.

File: arithmetization/src/test/java/net/consensys/linea/zktracer/precompiles/BlakeTests.java (hunk 6 above)
Comment (letypequividelespoubelles) [reply arithmetization/src/test/java/net/consensys/linea/zktracer/precompiles/BlakeTests.java]: oh yes, forget to make the change everywhere

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/imc/mmu/MmuCall.java (hunk 7 above)
Comment (letypequividelespoubelles) [reply arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/imc/mmu/MmuCall.java:594]: this one was easy ;)

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/blake2fmodexpdata/BlakeModexpDataOperation.java (hunk 8 above)
Comment (letypequividelespoubelles) [reply arithmetization/src/main/java/net/consensys/linea/zktracer/module/blake2fmodexpdata/BlakeModexpDataOperation.java:127]: just renaming. Everything was named input before (same name for callData, hashInput, and hashOutput ...)

File: arithmetization/src/test/java/net/consensys/linea/zktracer/precompiles/BlakeTests.java (hunk 6 above)
Comment (letypequividelespoubelles) [reply arithmetization/src/test/java/net/consensys/linea/zktracer/precompiles/BlakeTests.java]: done

**Pull request merged**